### PR TITLE
Refactor: Replace parts of boost::optional with std::optional (#6551,…

### PR DESCRIPTION
**Pull request by mugr1x**
# Issue

Since C++17 added support for 'containers', #6551 targets at replacing current code that uses boost::optional to use std::optional. This PR replaces some parts of the codebase with std::optional, indeed the ones I could convert without having to rewrite some parsers, see #6592 . I try complete these parser, but without that any further migrations to std::optional cause only myriads of compiler errors.

Please read our [documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/releasing.md) on release and version management.
If your PR is still work in progress please attach the relevant label.

## Tasklist

 - [X] CHANGELOG.md entry ([How to write a changelog entry](http://keepachangelog.com/en/1.0.0/#how))
 - [X] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki) ==> none edited
 - [X] add tests (see [testing documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/testing.md)) ==> no new functions to be tested
 - [ ] review
 - [ ] adjust for comments
 - [ ] cherry pick to release branch

## Requirements / Relations

#6551 
#6592

### CODE REVIEW

**Comment by mugr1x** [comment]
Tests fail with segfault.

**Comment by mugr1x** [comment]
Had something to do with DecreaseKey in query_heap.hpp

**Comment by mugr1x** [comment]
Relevant documentation includes:
https://www.boost.org/doc/libs/1_76_0/doc/html/boost/heap/d_ary_heap.html#id-1_3_18_6_2_1_1_1_22_17-bb

File: include/contractor/contract_excludable_graph.hpp [hunk 1]
```diff
@@ -61,10 +61,12 @@ inline auto contractExcludableGraph(ContractorGraph contractor_graph_,
         // Add all non-core edges to container
         {
             auto non_core_edges = toEdges<QueryEdge>(contractor_graph);
-            auto new_end =
-                std::remove_if(non_core_edges.begin(), non_core_edges.end(), [&](const auto &edge) {
-                    return is_shared_core[edge.source] && is_shared_core[edge.target];
-                });
+            auto new_end = std::remove_if(non_core_edges.begin(),
```
Comment (SiarheiFedartsou) [inline include/contractor/contract_excludable_graph.hpp:64]: Hm, do I get it correctly that most of changes in this PR are format changes. Were they pushed by mistake?

Comment (mugr1x) [reply include/contractor/contract_excludable_graph.hpp]: Better now ?

File: include/util/query_heap.hpp [hunk 2]
```diff
@@ -289,7 +289,7 @@ class QueryHeap
         return inserted_nodes[index].node == node;
     }
 
-    boost::optional<HeapNode &> GetHeapNodeIfWasInserted(const NodeID node)
+    std::optional<HeapNode> GetHeapNodeIfWasInserted(NodeID node)
```
Comment (SiarheiFedartsou) [inline include/util/query_heap.hpp:292]: I.e. now on each call we will copy HeapNode, right? How do we know that it won’t impact performance? May be worth refactoring it to `HeapNode*` instead(I mean use just pointer instead of optional<HeapNode> here)? From some perspective optional<T&> is just overcomplicated way to express simple pointer…

WDYT?


P.S. I didn’t have a time yet to look over the rest of PR, just something important what I noticed…

Comment (SiarheiFedartsou) [reply include/util/query_heap.hpp:292]: More than it: here you can see an example when we change contents of node in a heap and that by reference… https://github.com/Project-OSRM/osrm-backend/blob/d51631401eeab6e90fa58b29f2ba8c0f56dc1161/include/engine/routing_algorithms/routing_base_mld.hpp#L331

I.e. if we return it not by reference now most likely existing logic is broken.

Comment (mugr1x) [reply include/util/query_heap.hpp:292]: I did run my build and it did at least not segfault or got killed by kernel for too much RAM.
To cite from cppreference.com:
"There are no optional references; a program is ill-formed if it instantiates an optional with a reference type."
Search engine would suggest std::reference_wrapper for this task.

Comment (SiarheiFedartsou) [reply include/util/query_heap.hpp:292]: At the same time as I can see if we just change it to simple pointer instead of optional(I.e. won’t use neither boost::optional nor std::optional) the rest of code which uses this QueryHeap would “just work” without changes we had to make there…

Comment (mugr1x) [reply include/util/query_heap.hpp:292]: Regarding the other comment std::optional defines an operator->. However in some cases you probably would need to change a deref etc. I'm just testing solely using pointers for these parts. Does not work out of the box, but the idea should be doable.

Comment (SiarheiFedartsou) [reply include/util/query_heap.hpp:292]: `boost::optional<T&>` seems to be effectively the same as `T*`, isn’t it?

Comment (mugr1x) [reply include/util/query_heap.hpp:292]: Yes it is. Still forking the main branch I need replace some of the idioms. Do you prefer nullptr or some void coercion ?

Comment (SiarheiFedartsou) [reply include/util/query_heap.hpp:292]: What is void coercion? You mean `if (ptr == nullptr)` vs `if (!ptr)`? Personally I like first option as more explicit, but I don’t think it would be blocker to merge this PR :)

Comment (mugr1x) [reply include/util/query_heap.hpp]: Actually it's dreadfully simple, memory safety should be fine since it tests. I got too fed up with the nullptr version.
I think you need to approve CI again.

File: include/util/query_heap.hpp [hunk 3]
```diff
@@ -356,13 +356,13 @@ class QueryHeap
         const auto index = node_index.peek_index(node);
         auto &reference = inserted_nodes[index];
         reference.weight = weight;
-        heap.increase(reference.handle, std::make_pair(weight, index));
+        heap.decrease(reference.handle, std::make_pair(weight, index));
```
Comment (danpat) [inline include/util/query_heap.hpp:359]: This change is pretty fundamental to how the whole routing algorithm works.  Can you explain exactly why this change is being made here?

**Review by danpat — CHANGES_REQUESTED** [review]
Can you explain why the change from `heap.increase()` to `heap.decrease()` ?  "The old version was wrong because XYZ?" is an acceptable answer, but this is a pretty fundamental change to a core part of the routing algorithm, I think it needs a clear rationale.

**Comment by mugr1x** [comment]
@SiarheiFedartsou @danpat
Can anyone please approve CI again ?

**Comment by danpat** [comment]
@mugr1x You didn't address my question about the changes in `DecreaseKey` in `query_heap.hpp` - why are those changes included?  It's not obvious to me why changing to returning a pointer involves switching from `.increase()` to `.decrease()`.

**Comment by mugr1x** [comment]
Well I linked the docs two times and explained my reasoning. I opened a separate issue #6603 for this but I would really advise you too look into this anyway soonish. As said I fixed a segfault at some point for me.
Also you are completely right it has nothing to do with just returning a pointer in GetHeapNodeIfWasInserted, however this is the same in master, because as Siarhei pointed out boost::optional<T&> is just a disguise for T*.
So the function always returned a pointer nothing is changed by me except the idiom.

**Comment by mugr1x** [comment]
> > Well I linked the docs two times and explained my reasoning. I opened a separate issue #6603 for this but I would really advise you too look into this anyway soonish.
> 
> Can we discuss that in separate PR? This change is serious enough to deserve separate PR... Also if you see that it doesn't work in certain cases it is always a good practice to add test for such case to not regress it in the future.
Indeed

**Comment by mugr1x** [comment]
> > Well I linked the docs two times and explained my reasoning. I opened a separate issue #6603 for this but I would really advise you too look into this anyway soonish.
> 
> Can we discuss that in separate PR? This change is serious enough to deserve separate PR... Also if you see that it doesn't work in certain cases it is always a good practice to add test for such case to not regress it in the future.

Yes

**Comment by mugr1x** [comment]
I just noted that PR #6596 was merged, I had the same fix as that in this PR too. So I just resolved this one with master.

**Comment by mugr1x** [comment]
@danpat Are you happy with the current status ? If so can you close the review or similar.
Also CI fails with a weird error with redifinition of block_counter, which is clearly not in the source.
However I had the same error (with my own but literally equivalent fix as the one merged into master) sometimes.

**Comment by AlTimofeyev** [comment]
Hi @mugr1x , I took a glance at that error you mentioned and it looks like the file where `block_counter` is mentioned in your PR is redefined. Take a glance at your [commits](https://github.com/Project-OSRM/osrm-backend/pull/6593/commits/c42d32620eb97ed1ce637a9eb465d89fde011006). File `range_table.hpp` has two definitions for that variable on lines 84 and 86. It must've happened when you resolved with master (my bad about that).

**Comment by mugr1x** [comment]
@AlTimofeyev Yes, I first merged my own and your fix. You are right. I just did look at my local source not the the diff.
Thank you very much.

**Comment by mugr1x** [comment]
CI shows exactly the same error about block_counter, which is weird since it is clearly only defined once now, also in diff.

**Comment by SiarheiFedartsou** [comment]
@mugr1x may I ask you to create one more empty commit to trigger CI again? I suspect it is some bug in GitHub Actions.

File: include/util/range_table.hpp [hunk 4]
```diff
@@ -79,9 +79,9 @@ template <unsigned BLOCK_SIZE, storage::Ownership Ownership> class RangeTable
         unsigned last_length = 0;
         unsigned lengths_prefix_sum = 0;
         unsigned block_idx = 0;
-        unsigned block_counter = 0;
         BlockT block;
 #ifndef BOOST_ASSERT_IS_VOID
+        unsigned block_counter = 0;
```
Comment (danpat) [inline include/util/range_table.hpp:84]: These changes were merged to master in another PR https://github.com/Project-OSRM/osrm-backend/pull/6596 - they should not also be showing up in _this_ PR.  I suspect some merges have been incorrectly applied or lost on the branch.  This is probably also why CI is failing.

Comment (AlTimofeyev) [reply include/util/range_table.hpp:84]: The source code on master branch has `block_counter = 0` on line 85, meanwhile in the commit it is placed on line 84. Could that be what's causing CI to fail?

Comment (SiarheiFedartsou) [reply include/util/range_table.hpp:84]: @mugr1x may be try to squash all your commits into single one? Or open another PR? Tbh for me it is not clear why it is happening - looks more like some GH bug.
